File: Server/FakeFacebook/FakeFacebook/AES/AESEncryption.cs
using System.Text;

namespace FakeFacebook.AES
{
    public class AESEncryption
    {

        public static string Encryption(string StringToEncoded, string Key)
        {
            byte[] keyBytes = Encoding.UTF8.GetBytes(Key);
            byte[,] KeyExpansion = AESKeyExpansion.ExpandKey(keyBytes);
            byte[] StringBytes = Encoding.UTF8.GetBytes(StringToEncoded);

            var StringEncoded = "";

            int FullBlockCount = StringBytes.Length / 16;
            int RemainingBytes = StringBytes.Length % 16;

            //  Mã hóa các block đầy đủ
            for (int i = 0; i < FullBlockCount; i++)
            {
                byte[,] State4x4 = new byte[4, 4];
                for (int col = 0; col < 4; col++)
                {
                    for (int row = 0; row < 4; row++)
                    {
                        int index = i * 16 + (col * 4 + row);
                        State4x4[row, col] = StringBytes[index];
                    }
                }
                StringEncoded += BlockEncryption(State4x4, KeyExpansion);
            }

            //  Xử lý phần dư cuối hoặc trường hợp vừa đủ block
            byte[] LastBlock = null;

            if (RemainingBytes == 0)
            {
                LastBlock = new byte[16];
                for (int i = 0; i < 16; i++)
                {
                    LastBlock[i] = 16; // 0x10
                }
            }
            else
            {
                // Nếu còn dư byte, copy và thêm ký tự
                LastBlock = new byte[16];
                Array.Copy(StringBytes, FullBlockCount * 16, LastBlock, 0, RemainingBytes);
                byte paddingValue = (byte)(16 - RemainingBytes);
                for (int i = RemainingBytes; i < 16; i++)
                {
                    LastBlock[i] = paddingValue;
                }
            }

            // Mã hóa block cuối cùng
            byte[,] LastState4x4 = new byte[4, 4];
            for (int col = 0; col < 4; col++)
            {
                for (int row = 0; row < 4; row++)
                {
                    int index = col * 4 + row;
                    LastState4x4[row, col] = LastBlock[index];
                }
            }

            StringEncoded += BlockEncryption(LastState4x4, KeyExpansion);

            return StringEncoded;
        }

        // Mã hóa cho 1 block
        public static string BlockEncryption(byte[,] StringBytes, byte[,] KeyExpansion )
        {
            byte[,] FirstRound = new byte[4, 4];
            for (int col = 0; col < 4; col++)
            {
                FirstRound[0, col] = KeyExpansion[col, 0];
                FirstRound[1, col] = KeyExpansion[col, 1];
                FirstRound[2, col] = KeyExpansion[col, 2];
                FirstRound[3, col] = KeyExpansion[col, 3];
            }
            AddRoundKey(StringBytes, FirstRound);
            for(int rod=1;rod <=10; rod++)
            {
                byte[,] KeyRound = new byte[4, 4];
                for(int col = 0; col < 4; col++)
                {
                    int ColumnNumber = rod * 4 + col;
                    KeyRound[0, col] = KeyExpansion[ ColumnNumber, 0];
                    KeyRound[1, col] = KeyExpansion[ ColumnNumber,1];
                    KeyRound[2, col] = KeyExpansion[ ColumnNumber,2];
                    KeyRound[3, col] = KeyExpansion[ ColumnNumber,3];
                }
                SubBytes(StringBytes);
                ShiftRows(StringBytes);
                if (rod<10)
                {
                    MixColumns(StringBytes);
                }        
                AddRoundKey(StringBytes, KeyRound);

            }
            return GetHexFromState(StringBytes);
       

        }

        public static string GetHexFromState(byte[,] state)
        {
            StringBuilder hexString = new StringBuilder();
            int rows = state.GetLength(0);
            int cols = state.GetLength(1);

            for (int col = 0; col < cols; col++)
            {
                for (int row = 0; row < rows; row++)
                {
                    hexString.Append(state[row, col].ToString("X2"));
                }
            }

            return hexString.ToString();
        }


        // Hàm nhân 2 số trong GF(2^8)
        public static byte GFMul(byte a, byte b)
        {
            byte p = 0;
            for (int counter = 0; counter < 8; counter++)
            {
                if ((b & 1) != 0)
                {
                    p ^= a;
                }
                bool hiBitSet = (a & 0x80) != 0;
                a <<= 1;
                if (hiBitSet)
                {
                    a ^= 0x1B;
                }

                b >>= 1;
            }

            return p;
        }

        // Hàm MixColumns, đầu vào là ma trận state 4x4 bytes
        public static void MixColumns(byte[,] state)
        {
            for (int col = 0; col < 4; col++) // Lặp qua từng cột
            {
                byte s0 = state[0, col];
                byte s1 = state[1, col];
                byte s2 = state[2, col];
                byte s3 = state[3, col];

                // Tính toán theo ma trận MixColumns
                state[0, col] = (byte)(GFMul(0x02, s0) ^ GFMul(0x03, s1) ^ s2 ^ s3);
                state[1, col] = (byte)(s0 ^ GFMul(0x02, s1) ^ GFMul(0x03, s2) ^ s3);
                state[2, col] = (byte)(s0 ^ s1 ^ GFMul(0x02, s2) ^ GFMul(0x03, s3));
                state[3, col] = (byte)(GFMul(0x03, s0) ^ s1 ^ s2 ^ GFMul(0x02, s3));
            }
        }

        public static void SubBytes(byte[,] state)
        {
            for (int col = 0; col < 4; col++)
            {
                state[0, col] = SBoxTable.SubByte(state[0, col]);
                state[1, col] = SBoxTable.SubByte(state[1, col]);
                state[2, col] = SBoxTable.SubByte(state[2, col]);
                state[3, col] = SBoxTable.SubByte(state[3, col]);

            }

        }
        public static void ShiftRows(byte[,] state)
        {
            byte temp;

            // Hàng 1: dịch trái 1 lần
            temp = state[1, 0];
            state[1, 0] = state[1, 1];
            state[1, 1] = state[1, 2];
            state[1, 2] = state[1, 3];
            state[1, 3] = temp;

            // Hàng 2: dịch trái 2 lần (hoán đổi đối xứng)
            byte temp1 = state[2, 0];
            byte temp2 = state[2, 1];

            state[2, 0] = state[2, 2];
            state[2, 1] = state[2, 3];
            state[2, 2] = temp1;
            state[2, 3] = temp2;

            // Hàng 3: dịch trái 3 lần (hay dịch phải 1 lần)
            temp = state[3, 3];
            state[3, 3] = state[3, 2];
            state[3, 2] = state[3, 1];
            state[3, 1] = state[3, 0];
            state[3, 0] = temp;
        }

        public static void AddRoundKey(byte[,] state, byte[,] roundKey)
        {
            for (int row = 0; row < 4; row++) // duyệt từng hàng
            {
                for (int col = 0; col < 4; col++) // duyệt từng cột
                {
                    state[row, col] ^= roundKey[row, col]; // XOR giá trị state với khóa tương ứng
                }
            }
        }



    }
}

File: Client/FakeFacebook/src/app/AES/AESEncryption.ts

import { SBoxTable } from "./SBoxTable";
import { AESKeyExpansion } from "./AESKeyExpansion";
export class AESEncryption {
    static encryption(stringToEncode: string, key: string): string {
        const keyBytes = new TextEncoder().encode(key);
        const keyBytesArray: number[] = Array.from(keyBytes); // Chuyển đổi từ Uint8Array sang number[]
        const keyExpansion = AESKeyExpansion.expandKey(keyBytesArray);
        
        const stringBytes1 = new TextEncoder().encode(stringToEncode);
        const stringBytes: number[] = Array.from(stringBytes1); // Chuyển đổi từ Uint8Array sang number[]
        
        let stringEncoded = "";

        const fullBlockCount =  Math.floor(stringBytes.length / 16);
        const remainingBytes = stringBytes.length % 16;

        // Mã hóa các block đầy đủ


        for (let i = 0; i < fullBlockCount; i++) {
        let state4x4: number[][] = [];
        for (let col = 0; col < 4; col++) {
            state4x4[col] = [];  // Khởi tạo mảng con cho mỗi cột
            for (let row = 0; row < 4; row++) {
                const index = i * 16 + (col * 4 + row);
                state4x4[col][row] = stringBytes[index];  // Gán giá trị vào mảng con của cột
            }
        }
        stringEncoded += AESEncryption.blockEncryption(state4x4, keyExpansion);
        }
        

        // Xử lý phần dư cuối hoặc trường hợp vừa đủ block
        let lastBlock: number[] | null = null;

        if (remainingBytes === 0) {
            lastBlock = new Array(16).fill(16); // padding bằng 0x10
        } else {
        lastBlock = new Array(16).fill(0); // Tạo mảng 16 phần tử với giá trị mặc định là 0
        lastBlock.splice(0, remainingBytes, ...stringBytes.slice(fullBlockCount * 16, fullBlockCount * 16 + remainingBytes));
        
        // Thêm padding
        const paddingValue = 16 - remainingBytes;
        for (let i = remainingBytes; i < 16; i++) {
            lastBlock[i] = paddingValue;
        }
        }

        // Mã hóa block cuối cùng
        let lastState4x4: number[][] = Array.from({ length: 4 }, () => new Array(4).fill(0));
        for (let col = 0; col < 4; col++) {
            for (let row = 0; row < 4; row++) {
                const index = col * 4 + row;
                lastState4x4[row][col] = lastBlock[index];
            }
        }


        stringEncoded += AESEncryption.blockEncryption(lastState4x4, keyExpansion);



        return stringEncoded;
    }

    static blockEncryption(stringBytes: number[][], keyExpansion: number[][]): string {
        let firstRound: number[][] = Array.from({ length: 4 }, () => new Array(4).fill(0));

        for (let col = 0; col < 4; col++) {
            firstRound[0][col] = keyExpansion[col][0];
            firstRound[1][col] = keyExpansion[col][1];
            firstRound[2][col] = keyExpansion[col][2];
            firstRound[3][col] = keyExpansion[col][3];
        }
        

        AESEncryption.addRoundKey(stringBytes, firstRound);

        for (let round = 1; round <= 10; round++) {
            let keyRound: number[][] = Array.from({ length: 4 }, () => new Array(4).fill(0));
            for (let col = 0; col < 4; col++) {
                let columnNumber = round * 4 + col;
                keyRound[0][col] = keyExpansion[columnNumber][0];
                keyRound[1][col] = keyExpansion[columnNumber][1];
                keyRound[2][col] = keyExpansion[columnNumber][2];
                keyRound[3][col] = keyExpansion[columnNumber][3];
            }
            

            AESEncryption.subBytes(stringBytes);
            AESEncryption.shiftRows(stringBytes);

            if (round < 10) {
                AESEncryption.mixColumns(stringBytes);
            }

            AESEncryption.addRoundKey(stringBytes, keyRound);
        }

        return AESEncryption.getHexFromState(stringBytes);
    }

    static getHexFromState(state: number[][]): string {
        let hexString = '';
        for (let col = 0; col < 4; col++) {
            for (let row = 0; row < 4; row++) {
                hexString += state[row][col].toString(16).padStart(2, '0').toUpperCase();
            }
        }
        return hexString;
    }

    static gfMul(a: number, b: number): number {
        let p = 0;
        for (let counter = 0; counter < 8; counter++) {
            if ((b & 1) !== 0) {
                p ^= a;
            }
            const hiBitSet = (a & 0x80) !== 0;
            a <<= 1;
            if (hiBitSet) {
                a ^= 0x1B;
            }
            b >>= 1;
        }
        return p & 0xFF; // Đảm bảo kết quả nằm trong phạm vi byte
    }


    static mixColumns(state: number[][]): void {
        for (let col = 0; col < 4; col++) {
            let s0 = state[0][col];
            let s1 = state[1][col];
            let s2 = state[2][col];
            let s3 = state[3][col];

            state[0][col] = AESEncryption.gfMul(0x02, s0) ^ AESEncryption.gfMul(0x03, s1) ^ s2 ^ s3;
            state[1][col] = s0 ^ AESEncryption.gfMul(0x02, s1) ^ AESEncryption.gfMul(0x03, s2) ^ s3;
            state[2][col] = s0 ^ s1 ^ AESEncryption.gfMul(0x02, s2) ^ AESEncryption.gfMul(0x03, s3);
            state[3][col] = AESEncryption.gfMul(0x03, s0) ^ s1 ^ s2 ^ AESEncryption.gfMul(0x02, s3);
        }
    }

    static subBytes(state: number[][]): void {
        for (let col = 0; col < 4; col++) {
            state[0][col] = SBoxTable.subByte(state[0][col]);
            state[1][col] = SBoxTable.subByte(state[1][col]);
            state[2][col] = SBoxTable.subByte(state[2][col]);
            state[3][col] = SBoxTable.subByte(state[3][col]);
        }
    }

    static shiftRows(state: number[][]): void {
        let temp;
        // Row 1: shift left 1 time
        temp = state[1][0];
        state[1][0] = state[1][1];
        state[1][1] = state[1][2];
        state[1][2] = state[1][3];
        state[1][3] = temp;

        // Row 2: shift left 2 times
        const temp1 = state[2][0];
        const temp2 = state[2][1];

        state[2][0] = state[2][2];
        state[2][1] = state[2][3];
        state[2][2] = temp1;
        state[2][3] = temp2;

        // Row 3: shift left 3 times (or right 1 time)
        temp = state[3][3];
        state[3][3] = state[3][2];
        state[3][2] = state[3][1];
        state[3][1] = state[3][0];
        state[3][0] = temp;
    }

    static addRoundKey(state: number[][], roundKey: number[][]): void {
        for (let row = 0; row < 4; row++) {
            for (let col = 0; col < 4; col++) {
                state[row][col] ^= roundKey[row][col];
            }
        }
    }
}
